server/app/auth/oidc/OidcProvider.java

```diff
@@ -0,0 +1,185 @@
+
+    baseUrl = getConfigurationValue("base_url");
+  }
+
+  /*
+   * Provide the prefix used in the application.conf
+   */
+  protected abstract String attributePrefix();
+
+  /*
+   * Provide the profile adaptor that should be used.
+   */
+  public abstract ProfileCreator getProfileAdapter(OidcConfiguration config, OidcClient client);
+
+  protected String getConfigurationValue(String attr, String defaultValue) {
+    if (configuration.hasPath(attr)) {
+      return configuration.getString(attr);
+    }
+    return defaultValue;
+  }
+
+  protected String getConfigurationValue(String attr) {
+    return getConfigurationValue(attr, "");
+  }
+
+  protected String getProviderName() {
+    return getConfigurationValue(attributePrefix() + "." + providerNameConfigName);
+  }
+
+  protected String getClientID() {
+    return getConfigurationValue(attributePrefix() + "." + clientIDConfigName);
+  }
+
+  protected String getClientSecret() {
+    return getConfigurationValue(attributePrefix() + "." + clientSecretConfigName);
+  }
+
+  protected String getDiscoveryURI() {
+    return getConfigurationValue(attributePrefix() + "." + discoveryURIConfigName);
+  }
+
+  protected String getResponseMode() {
+    return getConfigurationValue(attributePrefix() + "." + responseModeConfigName);
+  }
+
+  protected String getResponseType() {
+    // Our local fake IDCS doesn't support 'token' auth.
+    if (baseUrl.contains("localhost:")) {
+      return "id_token";
+    }
+    return "id_token token";
+  }
+
+  protected String[] getExtraScopes() {
+    return getConfigurationValue(attributePrefix() + "." + extraScopesConfigName).split(" ");
+  }
+
+  protected String getCallbackURL() {
+    return baseUrl + "/callback";
+  }
+
+  protected String getScopes() {
+    // Scopes are the other things that we want from the OIDC endpoint
+    // (needs to also be configured on provider side).
+    String[] extraScopes = getExtraScopes();
+
+    ArrayList<String> allClaims = new ArrayList<>();
+    Collections.addAll(allClaims, defaultScopes);
+    Collections.addAll(allClaims, extraScopes);
+    return String.join(" ", allClaims);
+  }
+
+  @Override
+  public OidcClient get() {
+    String clientID = Objects.toString(getClientID(), "");
+    String clientSecret = Objects.toString(getClientSecret(), "");
+    String discoveryURI = Objects.toString(getDiscoveryURI(), "");
+    String responseMode = Objects.toString(getResponseMode(), "");
+    String responseType = Objects.toString(getResponseType(), "");
+    String callbackURL = Objects.toString(getCallbackURL(), "");
+    String providerName = Objects.toString(getProviderName(), ""); // optional
+    String scopes = Objects.toString(getScopes(), "");
+    if (clientID.isBlank()
+        || clientSecret.isBlank()
+        || discoveryURI.isBlank()
+        || responseMode.isBlank()
+        || responseType.isBlank()
+        || callbackURL.isBlank()
+        || baseUrl.isBlank()) {
```

Comment:
suggestion (non-blocking): alternatively could do

```suggestion
    var requiredAttributes = ImmutableList.of(
        clientID,
        clientSecret,
        discoveryURI,
        responseMode,
        responseType,
        callbackURL,
        baseUrl);
    if (requiredAttributes.stream().anyMatch(String::isBlank)) {
```